grpc/py-diatheke/diatheke/streams.py

```diff
@@ -0,0 +1,187 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright(2019) Cobalt Speech and Language Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http: // www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from diatheke_pb2 import AudioInput, ASRRequest
+from queue import Queue
+import threading
+
+
+class AudioInputStream(object):
+    """AudioInputStream represents an ASR stream for a session. Clients use
+    this stream to push audio to the server for recognition.
+    """
+
+    def __init__(self, client_stub, session_id):
+        """Creates a new stream using the given gRPC client stub and session ID.
+        Most users do not need to call this constructor directly, and should
+        instead use the client's stream_audio_input method to create this stream.
+        """
+
+        # The Python implementation of gRPC is a little different from other
+        # languages. Rather than having send and receive methods for the stream,
+        # request streams in Python take an iterator or generator function. This
+        # is a little inconvenient for how we want our SDK to be structured, so
+        # we use our own stream class (which is an iterator) in conjunction with
+        # some threading to make this work.
+
+        # Set up a queue to transfer data from the write function to the
+        # __next__ function.
+        self.data_queue = Queue(maxsize=1)
+        self.session_id = session_id
+        self._id_sent = False
+        self._is_finished = False
+
+        # Set up a thread to run the StreamAudioInput function with this class
```

Comment:
Could you please elaborate on this a little bit? I'm not sure why we need to use a separate thread here.